Tests/CoffeeAddictsTests/CoffeeAddictsTests.swift

```diff
@@ -0,0 +1,67 @@
+import XCTest
+import class Foundation.Bundle
+@testable import CoffeeAddictsLibrary
+
+final class CoffeeAddictsTests: XCTestCase {
```

Comment:
There a bunch of relevant test, I love the fact that you don't have individual test classes for each of the files, this made me think that you did some sort of tdd.

But you do test implementation details like helpers and validation, that is fine, on other way to do it was to use the actual entry point of your app (in this case the command). One benefit would have been that you could then vary the implementation and relay on the tests to make sure nothing is broken.

What is your experience with tests? When do you write tests and what type of tests do you usually write?

Reply:
> There a bunch of relevant test, I love the fact that you don't have individual test classes for each of the files, this made me think that you did some sort of tdd.
> 
> But you do test implementation details like helpers and validation, that is fine, on other way to do it was to use the actual entry point of your app (in this case the command). One benefit would have been that you could then vary the implementation and relay on the tests to make sure nothing is broken.
> 
> What is your experience with tests? When do you write tests and what type of tests do you usually write?

In this particular case of a Swift Package (my first package written from scratch to be honest), I just found out while implementing, that you cannot unit test an executable package, only a library one. It seems that it’s a limitation due to the fact that by default a package doesn’t have a project file. If you add one then you break the cross-platform functionality. Therefore I had to pack most of the code into a library and then make the “main app” to use it. 

So due to this fact that I spent hours on figuring this out (not much documentation available for this atm), I made a trade-off between how much time I should/can spend on implementing this project and to show that writing tests is important. 

That being said, I did have in mind lots of other tests to write: 
- test valid/invalid command line arguments and match the expected error messages
- test all the possible CSV file scenarios by importing local sample files (empty file, malformed line endings, extra attributes, invalid attributes order etc), test distance calculations

I used the TDD approach on a new project at some point, but the process was pretty slow mainly due to constant requirements changing (not only I had to constantly rewrite code but also maintaining already existing tests). And since the business requirement was to get the product on the market fast, I gave up on TDD.

Nowadays, since in the past two years I worked mostly on projects that have to ship fast, I write unit tests for complex calculations, for dates manipulation and date formatting, files manipulation, session storage etc. I write some integration tests for regression purpose, testing a whole end-to-end flow.